RSCBC-229: Observability updates (#11)
couchbase/docs-sdk-rust: modules/howtos/pages/observability-tracing.adoc

[abstract]
{description}

> The SDK instruments all operations with https://docs.rs/tracing/latest/tracing/struct.Span.html[`tracing`] spans.

Reviewer: Not sure that *instruments* is the clearest verb here - what do you think?

Reviewer 2: Instruments is the correct term within the observability context I think.

Reviewer: Okay - it feels a little jargon-y, but we'll put it through for now.  :+1:

tracing-subscriber = { version = "0.3", features = ["fmt"] }
----

The simplest setup registers `ThresholdLoggingTracer` together with a `fmt` layer for general log output.
This must be called within a Tokio runtime:

[source,rust]
----
include::devguide:example$src/observability.rs[tag=tracing-configure,indent=0]
----

The JSON blob emitted looks similar to the following (prettified here for readability):

[source,json]
----
{
   "kv": {
      "total_count": 2,
      "top_requests": [
         {
            "total_duration_us": 13333,
            "last_dispatch_duration_us": 13112,
            "total_dispatch_duration_us": 13112,
            "operation_name": "get",
            "last_local_id": "80d81672-0a71-4ab6-8667-80c0f16d3710",
            "last_remote_socket": "192.168.106.129:11210"
         },
         {
            "total_duration_us": 6252,
            "encode_duration_us": 10,
            "last_dispatch_duration_us": 6159,
            "total_dispatch_duration_us": 6159,
            "operation_name": "upsert",
            "last_local_id": "4e459355-9550-45fb-b5f4-ea71f5d4d613",
            "last_remote_socket": "192.168.106.130:11210"
         }
      ]
   }
}
----

> For each service (e.g. Key/Value or Query) an entry exists in the outer JSON array. The top N (10 by default) slowest operations are collected and displayed, sorted by the total duration. This promotes quick visibility of the "worst offenders" and more efficient troubleshooting.

Reviewer: ```suggestion
For each service (e.g. Key/Value or Query) an entry exists in the outer JSON array.
The top N (10 by default) slowest operations are collected and displayed, sorted by the total duration.
This promotes quick visibility of the "worst offenders" and more efficient troubleshooting.
```
No more than one sentence per line in docs source.

> **Note:** The stability of this output format is Uncommitted and may change between minor versions.

Reviewer: ```suggestion
**Note:** The stability of this output format is `Uncommitted` and may change between minor versions.
```

opentelemetry_sdk = "0.31"
opentelemetry-otlp = { version = "0.31", features = ["grpc-tonic"] }
----

Configure the OpenTelemetry pipeline and set up your OpenTelemetryLayer:

[source,rust]
----
include::devguide:example$src/observability.rs[tag=otel-tracing,indent=0]
----

NOTE: You can also register the subscriber with `tracing::subscriber::set_global_default(subscriber)`.
However, `tracing-opentelemetry` attaches a single `SdkTracer`, so all tracing spans in the process will share this tracer.

=== OpenTelemetry Troubleshooting

* There are many ways to export spans.
The example is exporting OpenTelemetry Protocol (OTLP) spans over GRPC to port 4317, which we believe is the _de facto_ standard for OpenTelemetry.
Make sure that your receiver is compatible with this, e.g. has these ports open and is ready to receive OTLP traffic over GRPC.
With https://www.jaegertracing.io/docs/1.41/getting-started/[Jaeger in Docker] this is achieved with the options `-e COLLECTOR_OTLP_ENABLED=true` and `-p 4317:4317`.
* The exporter used in this example is `BatchSpanProcessor`, which may not have a chance to export spans if the application exits very quickly (e.g. a test application).

> `SimpleSpanProcessor` can be used instead, though is not likely suitable for production.

Reviewer: ```suggestion
`SimpleSpanProcessor` can be used instead, though is not suitable for most production use cases.
```

* The example above uses `Sampler.alwaysOn()`, which exports every span.
This may need to be reduced to avoid overwhelming the receiver, with e.g. `Sampler.traceIdRatioBased(0.01)` to sample 1% of all traces.
* It can be worth sending traces into https://opentelemetry.io/docs/collector/[OpenTelemetry Collector], and forwarding them on from there to your receiver of choice.
Among other capabilities the collector can log traces it receives, making for easier debugging.

== Parent spans

> Because the SDK uses the `tracing` crate for instrumentation, parent span context is inherited automatically, any SDK call made while a `tracing` span is entered will nest its spans beneath it.

Reviewer: ```suggestion
Because the SDK uses the `tracing` crate for instrumentation, parent span context is inherited automatically.
Any SDK call made while a `tracing` span is entered will nest its spans beneath it.
```
